pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+	pub type IsOracleStopped<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn offchain_worker(block_number: T::BlockNumber) {
+			Self::fetch_price_from_oracle(block_number);
+		}
+	}
+
+	/// A public part of the pallet.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add `public` to the **list of Clamor Account IDs** that can ***validate*** and ***send*** **unsigned transactions with signed payload**
+		///
+		/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can edit this list
+		#[pallet::weight(10000)]
+		pub fn add_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("New key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.insert(public);
+			});
+
+			Ok(())
+		}
+
+		/// Remove a Clamor Account ID from `FragKeys`
+
+		/// NOTE: Only the Root User of the Clamor Blockchain (i.e the local node itself) can call this function
+		#[pallet::weight(10000)] // TODO
+		pub fn del_key(origin: OriginFor<T>, public: ed25519::Public) -> DispatchResult {
+			ensure_root(origin)?;
+
+			log::debug!("Removed key: {:?}", public);
+
+			<FragKeys<T>>::mutate(|validators| {
+				validators.remove(&public);
+			});
+
+			Ok(())
+		}
+
+		/// This function stores the new price received from the Oracle into a structure that contains a list of the last N-prices.
+		#[pallet::weight(10000)] // TODO
+		pub fn store_price(
+			origin: OriginFor<T>,
+			oracle_price: OraclePrice<T::Public, T::BlockNumber>,
+			_signature: T::Signature,
+		) -> DispatchResult {
+			// This ensures that the function can only be called via unsigned transaction.
+			ensure_none(origin)?;
+
+			log::debug!("Store price: {:?}", oracle_price);
+
+			let data_hash: H256 = oracle_price.using_encoded(blake2_256).into();
+
+			let latest_price: u32 =
+				oracle_price.price.try_into().map_err(|_| Error::<T>::SystematicFailure)?;
+			ensure!(!latest_price.is_zero(), Error::<T>::PriceIsZero);
+
+			let block_number = oracle_price.block_number;
+
+			// voting
+			let threshold = T::Threshold::get();
+			if threshold > 1 {
+				// 问Gio
+				let current_votes = <EVMLinkVoting<T>>::get(&data_hash);
+				if let Some(current_votes) = current_votes {
+					// Number of votes for the key `data_hash` in EVMLinkVoting
+					if current_votes + 1u64 < threshold {
+						// Current Votes has not passed the threshold
+						<EVMLinkVoting<T>>::insert(&data_hash, current_votes + 1);
+						return Ok(())
+					} else {
+						// Current votes passes the threshold, let's remove the record
+						<EVMLinkVoting<T>>::remove(&data_hash);
+					}
+				} else {
+					// If key `data_hash` doesn't exist in EVMLinkVoting
+					<EVMLinkVoting<T>>::insert(&data_hash, 1);
+					return Ok(())
+				}
+			}
+
+			log::trace!("Store new price: {}", latest_price);
+
+			<Price<T>>::put(latest_price);
```

Comment:
shouldn't this be `set`?

Reply:
Same as [other comment](https://github.com/fragcolor-xyz/clamor/pull/210#discussion_r1036669912).